sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2847,6 +2848,36 @@ class Analyzer(
     }
   }
 
+  object ResolveEncodersInUDF extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+      case p if !p.resolved => p // Skip unresolved nodes.
+
+      case p => p transformExpressionsUp {
+
+        case udf @ ScalaUDF(_, _, inputs, encoders, _, _, _) if encoders.nonEmpty =>
+          val resolvedEncoders = encoders.zipWithIndex.map { case (encOpt, i) =>
+            val dataType = inputs(i).dataType
+            if (dataType.isInstanceOf[UserDefinedType[_]]) {
+              // for UDT, we use `CatalystTypeConverters`
+              None
+            } else {
+              encOpt.map { enc =>
+                val attrs = if (enc.isSerializedAsStructForTopLevel) {
+                  dataType.asInstanceOf[StructType].toAttributes
+                } else {
+                  // the field name doesn't matter here, so we use
+                  // a simple literal to avoid any overhead
+                  new StructType().add("input", dataType).toAttributes
+                }
```

Comment:
Do we need to give attrs explicitly? Cannot the encoder use `schema` internally to resolve?

Reply:
By using internal `schema`, `ExpressionEncoder` would always use `Decimal(38, 18)` as the data type for the input decimal value when deserializing an internal row to external value. However, it could cause data loss when the data type of the input decimal value is not compatible with `Decimal(38, 18)`. For example, we should not cast the input value to `Deciaml(38, 18)` if the actual data type is `Decimal(30, 0)`. Therefore, we give the attrs explicitly here in order to let the encoder uses the actual data type from the input value when doing deserialzation.